crates/cli/src/commands/format.rs

```diff
@@ -0,0 +1,244 @@
+        .into_par_iter()
+        .map(|(file_path, resolved_patterns)| {
+            let result = format_file_resolved_patterns(&file_path, resolved_patterns, arg.clone());
+            (file_path, result)
+        })
+        .collect::<Vec<_>>();
+
+    // sort outputs to ensure consistent stdout output
+    // also avoid using sort_by_key to prevent additional cloning of file_path
+    results.sort_by(|(file_path, _), (other_file_path, _)| file_path.cmp(other_file_path));
+
+    for (file_path, result) in results {
+        match result {
+            Err(error) => eprintln!("couldn't format '{}': {error:?}", file_path),
+            Ok(Some(diff)) => println!("{}:\n{}", file_path.bold(), diff),
+            Ok(None) => (), // `args.write` is true or file is already formated
+        }
+    }
+    Ok(())
+}
+
+fn group_resolved_patterns_by_group(
+    resolved: Vec<ResolvedGritDefinition>,
+) -> Vec<(String, Vec<ResolvedGritDefinition>)> {
+    resolved.into_iter().fold(Vec::new(), |mut acc, resolved| {
+        let file_path = &resolved.config.path;
+        if let Some((_, resolved_patterns)) = acc
+            .iter_mut()
+            .find(|(resolv_file_path, _)| resolv_file_path == file_path)
+        {
+            resolved_patterns.push(resolved);
+        } else {
+            acc.push((file_path.clone(), vec![resolved]));
+        }
+        acc
+    })
+}
+
+fn format_file_resolved_patterns(
+    file_path: &str,
+    patterns: Vec<ResolvedGritDefinition>,
+    arg: FormatArgs,
+) -> Result<Option<DiffString>> {
+    let first_pattern = patterns
+        .first()
+        .ok_or_else(|| anyhow!("patterns is empty"))?;
+    let first_pattern_raw_data = first_pattern
+        .config
+        .raw
+        .as_ref()
+        .ok_or_else(|| anyhow!("pattern doesn't have raw data"))?;
+    let old_file_content = &first_pattern_raw_data.content;
+
+    let new_file_content = match first_pattern_raw_data.format {
+        PatternFileExt::Yaml => format_yaml_file(&patterns, old_file_content)?,
+        PatternFileExt::Grit => format_grit_code(old_file_content)?,
+        PatternFileExt::Md => {
+            let hunks = patterns
+                .iter()
+                .map(format_pattern_as_hunk_changes)
+                .collect::<Result<Vec<HunkChange>>>()?;
+            apply_hunk_changes(old_file_content, hunks)
+        }
+    };
+
+    if &new_file_content == old_file_content {
+        return Ok(None);
+    }
+
+    if arg.write {
+        std::fs::write(file_path, new_file_content).with_context(|| "could not write to file")?;
+        Ok(None)
+    } else {
+        Ok(Some(format_diff(old_file_content, &new_file_content)))
+    }
+}
+
+/// bubble clause that finds a grit pattern with name "\<pattern_name\>" in yaml and
+/// replaces it's body to "\<new_body\>", `format_yaml_file` uses this pattern to replace
+/// pattern bodies with formatted ones
+const YAML_REPLACE_BODY_PATERN: &str = r#"
```

Comment:
Composing a string then re-parsing it is going to be error prone / run into escaping issues.

Better to just build the pattern directly in Rust.

Reply:
> Composing a string then parsing it is going to be error prone / run into escaping issues.

Yeah, i was thinking about inserting variable to runtime and use it in the pattern, But couldn't find how to do it

> Better to just build the pattern directly in Rust.
That's gonna be hard :), Is there any place that i can see examples of this?

Reply:
> That's gonna be hard :), Is there any place that i can see examples of this?

Unfortunately don't have many examples today, you would have to build it up via the different pattern structs. I'm probably ok merging this as-is.

Reply:
> Unfortunately don't have many examples today, you would have to build it up via the different pattern structs.

What about inserting custom meta variable to runtime? Is it easy?

If not i think merging this is ok

Reply:
Yeah double quotes should be sufficient.